overlord/patch/patch6_2.go

```diff
@@ -0,0 +1,155 @@
+}
+
+type patch62Flags struct {
+	DevMode          bool `json:"devmode,omitempty"`
+	JailMode         bool `json:"jailmode,omitempty"`
+	Classic          bool `json:"classic,omitempty"`
+	TryMode          bool `json:"trymode,omitempty"`
+	Revert           bool `json:"revert,omitempty"`
+	RemoveSnapPath   bool `json:"remove-snap-path,omitempty"`
+	IgnoreValidation bool `json:"ignore-validation,omitempty"`
+	Required         bool `json:"required,omitempty"`
+	SkipConfigure    bool `json:"skip-configure,omitempty"`
+	Unaliased        bool `json:"unaliased,omitempty"`
+	Amend            bool `json:"amend,omitempty"`
+	IsAutoRefresh    bool `json:"is-auto-refresh,omitempty"`
+	NoReRefresh      bool `json:"no-rerefresh,omitempty"`
+	RequireTypeBase  bool `json:"require-base-type,omitempty"`
+}
+
+type patch62SnapState struct {
+	SnapType string             `json:"type"`
+	Sequence []*patch62SideInfo `json:"sequence"`
+	Active   bool               `json:"active,omitempty"`
+	Current  snap.Revision      `json:"current"`
+	Channel  string             `json:"channel,omitempty"`
+	patch62Flags
+	Aliases              interface{} `json:"aliases,omitempty"`
+	AutoAliasesDisabled  bool        `json:"auto-aliases-disabled,omitempty"`
+	AliasesPending       bool        `json:"aliases-pending,omitempty"`
+	UserID               int         `json:"user-id,omitempty"`
+	InstanceKey          string      `json:"instance-key,omitempty"`
+	CohortKey            string      `json:"cohort-key,omitempty"`
+	RefreshInhibitedTime *time.Time  `json:"refresh-inhibited-time,omitempty"`
+}
+
+type patch62SnapSetup struct {
+	Channel   string    `json:"channel,omitempty"`
+	UserID    int       `json:"user-id,omitempty"`
+	Base      string    `json:"base,omitempty"`
+	Type      snap.Type `json:"type,omitempty"`
+	PlugsOnly bool      `json:"plugs-only,omitempty"`
+	CohortKey string    `json:"cohort-key,omitempty"`
+	Prereq    []string  `json:"prereq,omitempty"`
+	patch62Flags
+	SnapPath     string           `json:"snap-path,omitempty"`
+	DownloadInfo interface{}      `json:"download-info,omitempty"`
+	SideInfo     *patch62SideInfo `json:"side-info,omitempty"`
+	patch62auxStoreInfo
+	InstanceKey string `json:"instance-key,omitempty"`
+}
+
+type patch62auxStoreInfo struct {
+	Media interface{} `json:"media,omitempty"`
+}
+
+// patch6_2:
+//  - ensure snapd snaps in the snapstate have TypeSnapd for backward compatibility with old snapd snap releases.
+//  - ensure snapd snaps have TypeSnapd in pending install tasks.
+func patch6_2(st *state.State) error {
+	var snaps map[string]*json.RawMessage
+	if err := st.Get("snaps", &snaps); err != nil && err != state.ErrNoState {
+		return fmt.Errorf("internal error: cannot get snaps: %s", err)
+	}
+
+	// migrate snapstate
+	for name, raw := range snaps {
+		var snapst patch62SnapState
+		if err := json.Unmarshal([]byte(*raw), &snapst); err != nil {
+			return err
+		}
+
+		if snapst.SnapType == string(snap.TypeSnapd) {
```

Comment:
this case doesn't seem to be exercised

Reply:
I've split this loop into two to make testing of these two cases possibles (this one and the one below).